lib/hobbes/lang/preds/dependent.C

```diff
@@ -0,0 +1,588 @@
+
+#include <hobbes/lang/expr.H>
+#include <hobbes/lang/preds/dependent.H>
+#include <hobbes/eval/funcdefs.H>
+
+namespace hobbes{
+
+static bool tvalueToMonoTypePtr(const TEnvPtr& tenv, Definitions* ds, const MonoTypePtr& in, MonoTypePtr& out) {    
+  if (is<TString>(in)) {
+    out = arrayty(primty("char"));
+    return true;
+  }      
+  if (is<TLong>(in)) {
+    out = primty("long");
+    return true;
+  }
+  if (TApp *tapp = is<TApp>(in)) {
+    if (*tapp->fn() == *primty("quote")) {
+      if (TExpr* e = is<TExpr>(tapp->args()[0])) {
+        out = validateType(tenv, e->expr(), ds)->type()->monoType();
+        return true;
+      }
+    }
+  }
+  return false; 
+}
+
+
+struct MonoTypePtrToExpr : public switchType<ExprPtr> {
+  TEnvPtr tenv;
+  MonoTypePtr type;
+  LexicalAnnotation la;
+  MonoTypePtrToExpr(const TEnvPtr tenv, const MonoTypePtr& type) : tenv(tenv), type(type) {}
+
+  ExprPtr roll(ExprPtr e) const {
+    return fncall(var("roll", la), list(e), la);
+  }
+
+  ExprPtr ret(const char * name, ExprPtr e) const {
+    ExprPtr o = fncall(var("roll", la), list(ExprPtr(new MkVariant(name, e, la))), la);
+    return ExprPtr(new Assump(o, qualtype(type), la));;
+  }
+
+  ExprPtr with(const TVar* tv) const {
+    return ret("TVar", ExprPtr(mkarray(tv->name(), la)));
+  }
+
+  ExprPtr with(const TGen*) const {
+    // I don't think we can get one of these here, and we don't want it.
+    throw std::runtime_error("TGen in TypeApply");
+  }
+
+  ExprPtr with(const TExpr *) const {
+    // We can't put an arbitrary value in a Type, can we do something else?
```

Comment:
Although we could just return the ExprPtr out of this TExpr, to make this reflection total we'll have to do the same thing for expr (and/or just bring the two types into one).